core/src/main/java/com/exadel/aem/backpack/core/servlets/UploadPackageServlet.java

```diff
@@ -0,0 +1,128 @@
+                "sling.servlet.paths=/services/backpack/uploadPackage",
+                "sling.servlet.methods=[get,post]"
+        })
+@SuppressWarnings("PackageAccessibility")
+// because Servlet and HttpServletResponse classes reported as a non-bundle dependency
+public class UploadPackageServlet extends SlingAllMethodsServlet {
+
+    private static final String PARAM_FORCE_UPDATE = "forceUpdate";
+    private static final String PARAM_FILEUPLOAD = "fileupload";
+    private static final long serialVersionUID = 1L;
+    private static final Gson GSON = new Gson();
+
+    @Reference
+    @SuppressWarnings("UnusedDeclaration") // value injected by Sling
+    private transient PackageService packageService;
+
+    /**
+     * Processes {@code POST} requests to the current endpoint. Attempts to create a package for a consequential build
+     * according to the request parameters.
+     * Request parameters are parsed to a {@link PackageModel} which is validated and passed
+     * to the corresponding {@link PackageService} routine if proven valid; otherwise, the {@code HTTP status 400} reported
+     *
+     * @param request  {@code SlingHttpServletRequest} instance
+     * @param response {@code SlingHttpServletResponse} instance
+     * @throws IOException in case writing data to the {@code SlingHttpServletResponse} fails
+     */
+    @Override
+    protected void doPost(final SlingHttpServletRequest request,
+                          final SlingHttpServletResponse response) throws IOException {
+        Session session = request.getResourceResolver().adaptTo(Session.class);
+        byte[] fileUploadBytesArray = getFileUploadBytesArray(request);
+        boolean forceUpdate = getForceUpdate(request);
+
+        JcrPackageWrapper jcrPackageWrapper = packageService.uploadPackage(session, fileUploadBytesArray, forceUpdate);
+
+        writeResponse(response, jcrPackageWrapper);
+    }
+
+    private byte[] getFileUploadBytesArray(final SlingHttpServletRequest request) {
+        String parameter = request.getParameter(PARAM_FILEUPLOAD);
+        if (parameter != null) {
+            return parameter.getBytes(ISO_8859_1);
+        }
+        return null;
+    }
+
+    private void writeResponse(final SlingHttpServletResponse response, final JcrPackageWrapper jcrPackageWrapper) throws IOException {
+        response.setContentType(APPLICATION_JSON);
+        response.setCharacterEncoding(UTF_8.name());
+        if (jcrPackageWrapper.isExist()) {
+            PackageInfo packageInfo = packageService.getPackageInfo(jcrPackageWrapper.getJcrPackage());
+            if (packageInfo != null) {
+                writeSuccess(response, packageInfo);
+                return;
+            } else {
+                jcrPackageWrapper.setStatusCode(SC_CONFLICT);
+                jcrPackageWrapper.setMessage("Something went wrong! Please ask administrator for assistance.");
+            }
+        }
+
+        writeError(response, jcrPackageWrapper);
```

Comment:
Write error, in the end, doesn't look logical, looks like we always write an error. What will happen in the case if we have an empty jcrPackageWrapper?

Reply:
From default "/crx/packmgr/" we haven't opportunity to download empty package. 
Uploading an empty package doesn't make any sens and should lead for exception.